### PR TITLE
[JENKINS-29870] Make Android Lint Plugin compatible with Workflow

### PR DESCRIPTION
**Pull request by recena**
Downstream of https://github.com/jenkinsci/analysis-core-plugin/pull/31

Make Android Lint Plugin compatible with Workflow [JENKINS-29870](https://issues.jenkins-ci.org/browse/JENKINS-29870)

```
node {
  checkout changelog: false, poll: false, scm: [$class: 'SubversionSCM', additionalCredentials: [], excludedCommitMessages: '', excludedRegions: '', excludedRevprop: '', excludedUsers: '', filterChangelog: false, ignoreDirPropChanges: false, includedRegions: '', locations: [[credentialsId: 'efaa090d-5627-4a5c-95df-f570850d89a1', depthOption: 'infinity', ignoreExternalsOption: true, local: '.', remote: 'https://server/svn/repo/trunk']], workspaceUpdater: [$class: 'UpdateUpdater']]
  sh 'mvn clean install -Dmaven.test.skip'
  step([$class: 'LintPublisher'])
}
```

- [x] Initial source code modifications
- [x] Add automated tests specific for Workflow compatibility

### CODE REVIEW

**Comment by recena** [comment]
@orrc your feedback would be great!

**Comment by recena** [comment]
@reviewbybees Could your review these changes?

File: src/main/java/org/jenkinsci/plugins/android_lint/LintResult.java [hunk 1]
```diff
@@ -27,24 +26,64 @@
...
+     * @param usePreviousBuildAsReference the value of usePreviousBuildAsReference
+     * @param useStableBuildAsReference determines whether only stable builds should be used as reference builds or not
+     * @param actionType the type of the result action
+     */
+    protected LintResult(final Run<?, ?> build,
+                        final String defaultEncoding, final ParserResult result,
+                        final boolean usePreviousBuildAsReference,
+                        final boolean useStableBuildAsReference,
+                        final Class<? extends ResultAction<LintResult>> actionType) {
+        super(build, new BuildHistory(build, actionType, usePreviousBuildAsReference, useStableBuildAsReference), result, defaultEncoding);
```
Comment (tfennelly) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintResult.java]: `super` and not `this`?

**Comment by tfennelly** [comment]
:bee: assuming `super`/`this` as pointed to above is not an issue.

NIT: I think it would be good for all of us to add more comments in tests (e.g. the `LintWorkflowTest` in this PR). I think the tests are a really useful source of knowledge, but only if they have a few comments to allow people make sense of them.

**Comment by amuniz** [comment]
:bee: as I don't see bugs, but I have two suggestions:

1. This plugin master branch is currently depending on a pretty old version of ```analysis-core-plugin```, so this PR is being a big jump. I think that at least a manual test verifying that old reports are loaded/shown correctly would be great. This is really an :ant:.
1. There are some missing features (related to ```usePreviousBuildAsReference ``` and ```useStableBuildAsReference ``` and probably more), adding this features (coming from ```analysis-core-plugin```) are clearly out of the scope of this PR, so it's your call include it or not. I would suggest to do it.

**Comment by recena** [comment]
@tfennelly @amuniz Thanks for your feedback. I'll back soon with reviews done.

**Comment by recena** [comment]
I totally agree with @amuniz in his first comment. I'm going to do some smoke tests before to include that features (`usePreviousBuildAsReference` and `useStableBuildAsReference`).

File: src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java [hunk 2]
```diff
@@ -15,9 +18,28 @@
...
+        this((Run<?, ?>) build, defaultEncoding, result, false, false);
+    }
+
+    /**
+     * @param build The current build as owner of this action.
+     * @param defaultEncoding The default encoding to be used when reading files.
+     * @param result The parsed result with all annotations.
+     */
+    @Deprecated
+    public LintReporterResult(final Run<?, ?> build, final String defaultEncoding, final ParserResult result) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java]: :ant: why are you introducing a member and deprecating it in the same PR?

**Comment by jglick** [comment]
:bee:

**Comment by recena** [comment]
@reviewbybees done

File: src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java [hunk 3]
```diff
@@ -18,20 +16,29 @@
...
     private static final String DEFAULT_PATTERN = "**/lint-results.xml";
 
-    private static final long serialVersionUID = 3435696173660003622L;
-
-    /** Ant fileset pattern of files to work with. */
-    private final String pattern;
+    /**
+     * Ant fileset pattern of files to work with.
+     */
+    private String pattern;
```
Comment (orrc) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java]: Most of the stuff above here seems to have been unnecessary formatting changes.

File: src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java [hunk 4]
```diff
@@ -88,6 +95,15 @@ public LintPublisher(final String healthy, final String unHealthy, final String
     }
 
     /**
+     * Constructor used from methods like {@link StaplerRequest#bindJSON(Class, JSONObject)} and
+     * {@link StaplerRequest#bindParameters(Class, String)}.
+     */
+    @DataBoundConstructor
+    public LintPublisher() {
```
Comment (orrc) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java]: This Javadoc doesn't seem helpful. I'm sure we can do without it.

File: src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java [hunk 5]
```diff
@@ -96,23 +112,32 @@ public String getPattern() {
         return pattern;
     }
 
+    /**
+     * Sets the Ant file-set pattern of files to work with.
+     */
+    @DataBoundSetter
+    public void setPattern(final String pattern) {
+        this.pattern = pattern;
+    }
```
Comment (orrc) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java]: Is there is a reason to add this method, rather than just setting `@DataBoundSetter` on the `pattern` field?
Is it required for setting options via Workflow?

File: src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java (hunk 2 above)
Comment (orrc) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java]: Yeah, why? :)

File: src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java [hunk 6]
```diff
@@ -15,9 +18,28 @@
...
+        this(build, defaultEncoding, result, false, false);
+    }
+
+    /**
+     * @param build The current build as owner of this action.
+     * @param defaultEncoding The default encoding to be used when reading files.
+     * @param result The parsed result with all annotations.
+     */
+    public LintReporterResult(final Run<?, ?> build, final String defaultEncoding, final ParserResult result, final boolean usePreviousBuildAsReference, final boolean useStableBuildAsReference) {
```
Comment (orrc) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java]: Javadoc is copy/pasted from above, ignoring the extra parameters.

File: src/test/java/org/jenkinsci/plugins/android_lint/LintResultTest.java [hunk 7]
```diff
@@ -25,7 +25,7 @@
     public void testResultDuplicates() throws ExecutionException, InterruptedException, IOException {
         // Create a project with a Lint results file in its workspace.
         FreeStyleProject project = jenkins.createFreeStyleProject();
-        project.setScm(new ExtractResourceSCM(getClass().getResource("duplicates.zip")));
+        project.setScm(new ExtractResourceSCM(getClass().getResource("parser/duplicates.zip")));
```
Comment (orrc) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintResultTest.java]: Since you've moved this test class to the correct package, you may as well move the dependent resource as well.

File: src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    /**
+     * Run a workflow job using {@link LintPublisher} and check for success.
+     */
+    @Test
+    public void lintPublisherWorkflowStep() throws Exception {
+        WorkflowJob job = jenkinsRule.jenkins.createProject(WorkflowJob.class, "lintPublisherWorkflowStep");
+        FilePath workspace = jenkinsRule.jenkins.getWorkspaceFor(job);
+        FilePath report = workspace.child("target").child("lint-results.xml");
+        report.copyFrom(LintWorkflowTest.class.getResourceAsStream("/org/jenkinsci/plugins/android_lint/parser/lint-results_r20.xml"));
```
Comment (orrc) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java]: This would be simpler as `report.copyFrom(getClass().getResourceAsStream("parser/lint-results_r20.xml"))`?

File: src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     * Run a workflow job using {@link LintPublisher} with a failing threshold of 0, so the given example file
+     * "/org/jenkinsci/plugins/android_lint/parser/lint-results_r20.xml" will make the build to fail.
+     */
+    @Test
+    public void lintPublisherWorkflowStepSetLimits() throws Exception {
+        WorkflowJob job = jenkinsRule.jenkins.createProject(WorkflowJob.class, "lintPublisherWorkflowStepSetLimits");
+        FilePath workspace = jenkinsRule.jenkins.getWorkspaceFor(job);
+        FilePath report = workspace.child("target").child("lint-results.xml");
+        report.copyFrom(LintWorkflowTest.class.getResourceAsStream("/org/jenkinsci/plugins/android_lint/parser/lint-results_r20.xml"));
```
Comment (orrc) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java]: Ditto.

File: src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     * Run a workflow job using {@link LintPublisher} with a unstable threshold of 0, so the given example file
+     * "/org/jenkinsci/plugins/android_lint/parser/lint-results_r20.xml" will make the build to fail.
+     */
+    @Test
+    public void lintPublisherWorkflowStepFailure() throws Exception {
+        WorkflowJob job = jenkinsRule.jenkins.createProject(WorkflowJob.class, "lintPublisherWorkflowStepFailure");
+        FilePath workspace = jenkinsRule.jenkins.getWorkspaceFor(job);
+        FilePath report = workspace.child("target").child("lint-results.xml");
+        report.copyFrom(LintWorkflowTest.class.getResourceAsStream("/org/jenkinsci/plugins/android_lint/parser/lint-results_r20.xml"));
```
Comment (orrc) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java]: Ditto.

File: src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        job.setDefinition(new CpsFlowDefinition(""
+                        + "node {\n"
+                        + "  step([$class: 'LintPublisher', pattern: '**/lint-results.xml', unstableTotalAll: '0', usePreviousBuildAsReference: false])\n"
```
Comment (orrc) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java]: There should probably be another test that uses something other than the default `**/lint-results.xml` value, otherwise you're never actually verifying that the `pattern` parameter works?

File: src/test/resources/org/jenkinsci/plugins/android_lint/parser/lint-results_r20.xml [hunk 12]
```diff
@@ -17,7 +17,7 @@
     <issue
         id="Registered"
         severity="Warning"
-        message="The &lt;activity> MonitoredActivity is not registered in the manifest">
+        message="The &lt;activity&gt; MonitoredActivity is not registered in the manifest">
```
Comment (orrc) [inline src/test/resources/org/jenkinsci/plugins/android_lint/parser/lint-results_r20.xml]: Why would you alter the existing test data?

Please revert this and the other XML test data change.

File: src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java (hunk 5 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java]: It is preferable to use `@DataBoundSetter` on a, well, setter :-) rather than on a field, for the usual reasons: better encapsulation, and makes future API changes easier to do compatibly.

**Comment by orrc** [comment]
Thanks for the PR. I gave the plugin a quick freestyle job test with the newer analysis-core version and from a brief look, it seems like it still works...

**Comment by recena** [comment]
@orrc All your comments will be addressed as soon as possible.

File: src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java (hunk 4 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java]: @orrc Removed.

File: src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java (hunk 3 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java]: @orrc Back the original format. I'd prefer my proposal:

```

/**
 * Publishes the results of parsing an Android lint file.
 */
public class LintPublisher extends HealthAwarePublisher {

    private static final long serialVersionUID = 3435696173660003622L;

    /**
     * Plugin name used in console output.
     */
    private static final String PLUGIN_NAME = "android-lint";

    /**
     * Default filename pattern
     */
    private static final String DEFAULT_PATTERN = "**/lint-results.xml";

    /**
     * Ant fileset pattern of files to work with.
     */
    private String pattern;
```

File: src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java (hunk 5 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java]: @jglick Thanks for your explanation.

File: src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java (hunk 2 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java]: @orrc I had an explanation but I've chosen a simplest way.

File: src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java (hunk 6 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintReporterResult.java]: @orrc My mistake. Reviewed.

File: src/test/java/org/jenkinsci/plugins/android_lint/LintResultTest.java (hunk 7 above)
Comment (recena) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintResultTest.java]: @orrc I did this changes in order to apply the same criteria that in other analysis plugins. Reviewed.

File: src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java (hunk 8 above)
Comment (recena) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java]: @orrc Done

**Comment by recena** [comment]
@orrc 

* I had to change line separators in `LintMavenResultAction` because there were mixed.
* It would be helpful if we remove the Eclipse project files & folder. This kind of data is created automatically.
* I have pending to review the changes in XMLs. There is a reason for that.

**Comment by recena** [comment]
@orrc Please, consider this PR for merge if you think everything is fine. Thanks so much for your review.

File: src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java (hunk 3 above)
Comment (orrc) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java]: > Back the original format

Thanks.

> I'd prefer my proposal

And I prefer the style which doesn't waste two lines of space for no reason whatsoever! :)

File: src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        FilePath report = workspace.child("target").child("lint-results.xml");
+        report.copyFrom(getClass().getResourceAsStream("./parser/lint-results_r20.xml"));
+        job.setDefinition(new CpsFlowDefinition(""
+                        + "node {\n"
+                        + "  step([$class: 'LintPublisher', pattern: '**/lint-results.xml', failedTotalAll: '0', usePreviousBuildAsReference: false])\n"
+                        + "}\n", true)
+        );
+        jenkinsRule.assertBuildStatus(Result.FAILURE, job.scheduleBuild2(0).get());
+        LintResultAction result = job.getLastBuild().getAction(LintResultAction.class);
+        assertTrue(result.getResult().getAnnotations().size() == 4);
```
Comment (orrc) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java]: Could you please use `assertEquals(4, ...)` instead of `assertTrue`?

File: src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java (hunk 11 above)
Comment (orrc) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java]: AFAICT, this is still the case?

The filename pattern defaults to `**/lint-results.xml` if no data-bound `pattern` value is provided, so here you're not actually testing whether the `pattern` parameter works.

i.e. even if you remove the body of the [`setPattern`](https://github.com/recena/android-lint-plugin/blob/831e3c0/src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java#L110) method, I believe these new workflow tests would still pass.

**Comment by orrc** [comment]
> I had to change line separators in LintMavenResultAction because there were mixed.

No, you really didn't have to :)  Sorry, but I'm not committing a patch with a crazy amount of whitespace changes.  I'm sure some files may be messed up, but if you really must fix the whitespace, please do it in a separate PR.

> It would be helpful if we remove the Eclipse project files & folder. This kind of data is created automatically.

I don't think it makes a difference, but sure — I'm not sure why I would have added them in the first place. But I can do that in a separate commit.

---

I'll try and review the updates this week — `git am` is rejecting the patch at the moment due to the whitespace change...

File: src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java (hunk 13 above)
Comment (recena) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java]: @orrc I don't know the reason but I did it.

File: src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java (hunk 3 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/android_lint/LintPublisher.java]: @orrc There are reasons but maybe you don't agree with them ;) It is your code, you decide.

**Comment by recena** [comment]
@orrc When you compare branches you can ignore changes related to end of lines. This is not a problem.

**Comment by recena** [comment]
@orrc I believe that all your comments have been addressed.

File: src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java (hunk 13 above)
Comment (orrc) [inline src/test/java/org/jenkinsci/plugins/android_lint/LintWorkflowTest.java]: Thanks. The reason is because `4` isn't a `boolean`, so `assertEquals` doesn't make sense and because using the correct method gives this:

```
java.lang.AssertionError: expected:<4> but was:<0>
        at org.junit.Assert.assertEquals(Assert.java:542)
        at ...(LintWorkflowTest.java:34)
```

Which is a far more helpful error message than:

```
java.lang.AssertionError: null
        at org.junit.Assert.assertTrue(Assert.java:52)
        at ...(LintWorkflowTest.java:34)
```

**Comment by orrc** [comment]
Looks good, thanks. If you could squash, that would be great.

**Comment by recena** [comment]
@orrc Thanks so much for your time.

**Comment by amuniz** [comment]
:bee:

**Comment by recena** [comment]
@orrc Thanks so much for your time. Remember update the `analysis-core` dependency when is released.
